### PR TITLE
module to generate RSA or ECDSA certificates

### PR DESCRIPTION
**Pull request by ghost**
##### Issue Type:
- New Module Pull Request
##### Plugin Name:

generate_ssl_cert
##### Ansible Version:

```
2.2.0
```
##### Summary:

This module is a wrapper around the openssl req command. It can be used to generate self-signed RSA or ECDSA certificates or certificate signing requests (CSRs).

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @aeibrahim for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

File: web_infrastructure/generate_ssl_cert.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+    description:
+      - Optionally provide a password to use to encrypt the generated private key. Does not log the provided password. If I(encrypt) is set to C(yes), you must specify a password.
+    required: false
+    default: null
+  type:
+    description:
+      - the type of certificate to generate. If type is C(ecdsa), you must specify a I(curve) to use.
+    required: false
+    default: "rsa"
+    choices: [ "rsa", "ecdsa" ]
```
Comment (gundalow) [inline web_infrastructure/generate_ssl_cert.py:65]: Could you please add:
`aliases = [ 'cert_type' ]`

File: web_infrastructure/generate_ssl_cert.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+    aliases: ['CN']
+  email:
+    description:
+      - The email portion of the subject in the certificate. This should be a valid email address to contain the organization. Usually the email address of the certificate administrator or IT department. 
+    required: false
+    default: null
+    aliases: ['email_address']
+  owner:
+    description:
+      - the owner id of the certificate and private key files, after execution.
```
Comment (gundalow) [inline web_infrastructure/generate_ssl_cert.py]: The...

File: web_infrastructure/generate_ssl_cert.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    default: null
+    aliases: ['email_address']
+  owner:
+    description:
+      - the owner id of the certificate and private key files, after execution. 
+    required: false
+    default: null
+  group:
+    description:
+      - the group id of the certificate and private key files, after execution.
```
Comment (gundalow) [inline web_infrastructure/generate_ssl_cert.py]: The...

File: web_infrastructure/generate_ssl_cert.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+    required: false
+    default: null
+  group:
+    description:
+      - the group id of the certificate and private key files, after execution. 
+    required: false
+    default: null
+  mode:
+    description:
+      - the permissions of the certificate file, after execution. The permissions of the private key is always set to 0600.
```
Comment (gundalow) [inline web_infrastructure/generate_ssl_cert.py]: The...

**Comment by gundalow** [comment]
Hi,
Thanks for the new module

I've only done a quick test so far however what I've tested has worked as expected. 
- I see why this module has been created. Using command and creates (to make it idempotent) wouldn't be clean
  - Sensible error checking, in particular giving different meaningful error messages for permission errors. Also dealing with openssl not being available
- Solid set of real world examples

1) Although not a hard requirement it would be good if the code was PEP8 clean. It's a lot easier for this to be done as part of the initial module creation. If this was to be fixed after the module is submitted it makes tracing where errors got introduced a little harder.

The Makefile in ansible/ansible defines the PEP8 standard to be

`pep8 -r --ignore=E501,E221,W291,W391,E302,E251,E203,W293,E231,E303,E201,E225,E261,E241,E402 generate_ssl_cert.py`

2) Similarly their is a fair bit of trailing whitespace
sed 's/[ \t]*$//' generate_ssl_cert.py

**Comment by gundalow** [comment]
needs_revision

**Comment by ghost** [comment]
Thanks @bob-smith for the feedback. I've corrected the issues you've raised. If there's anything else that needs updating, please do let me know.

**Comment by aeibrahim** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @aeibrahim for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by ghost** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @aeibrahim for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by evgkrsk** [comment]
shipit

**Comment by raben2** [comment]
shipit

**Comment by ghost** [comment]
I am adding additional functionality to this module, so it is still a work in progress. I'm finalizing the ability to create certificates with subject alternative names:

```
- name: generate_ssl_cert | create a CSR containing SANs
  generate_ssl_cert:
    dest: /tmp/example.com.csr
    self_signed: False
    country: CA
    state: Ontario
    city: Toronto
    org: ACME Corp.
    org_unit: Ansible Divsion
    common_name: "*.example.com"
    subjectAltName:
     - IP: 127.0.0.1
     - DNS: "www.example.com"
```

**Comment by gundalow** [comment]
Hi

Can @aeibrahim @Spredzy please have a look at the different PRs that are in flight that do similar things and between yourselves please work out which you think should go forward:

aeibrahim's: module to generate RSA or ECDSA certificates #1798 
Spredzy's: openssl_privatekey #2326, openssl_csr #2341 and openssl_cert #2343

**Comment by ghost** [comment]
Closed in favour of #2343.

**Comment by gundalow** [comment]
Thank you for reviewing this.

**Pull request closed without merging**